### PR TITLE
test(stress): prestart and retain fixed experiment workers

### LINKED ISSUE #3276: test(stress): prestart and retain fixed experiment workers
The maintained outbox run [34627120888](https://github.com/thomhurst/Dekaf/actions/runs/34627120888/job/103355117845) completed its workloads but was INCONCLUSIVE because worker counts changed during warmup/measurement, despite `fixed_worker_threads=32`.

The current workflow sets the .NET minimum and maximum worker limits to 32. Those limits do not prestart workers or prevent idle retirement. A controlled Linux experiment using the same SDK/runtime (10.0.401 / 10.0.12) confirms this: after explicitly starting 32 blocked callbacks and releasing them, all 32 workers retire within 25 seconds while GetMinThreads/GetMaxThreads still report 32/32. With `DOTNET_ThreadPool_ThreadTimeoutMs=-1`, all 32 remain after the same idle interval. Runtime decompilation shows the default 20-second timeout and an exit path that does not retain the minimum number of existing threads.

Exact source, test executable, Linux runtime files/image identity, raw outputs and runtime decompilations are retained outside worktrees at `D:/Dekaf-evidence/issue-3232/threadpool-probe`. The initial configuration-only probe archive is explicitly distinguished from the complete later idle experiment; its overwritten original executable is not claimed as retained.

Acceptance:

- Make the existing explicitly requested fixed-worker experiment prestart its configured worker count before workload warmup and retain those workers through idle phases.
- Apply the same configuration to both pinned controls and candidate. Keep the default input 0 and ordinary workload behavior unchanged.
- Verify actual runtime minimum, maximum and live worker count, record them in provenance, and fail before a paid workload if requested configuration cannot be established. Do not weaken the steady-state validator or discard changing samples.
- Add deterministic configuration/error coverage and a bounded process-level worker lifecycle check. Avoid changing the test runner's shared thread pool. Keep setup work outside measured per-message paths.
- Preserve all workload correctness, shutdown, phase duration and warmup contracts. Keep diagnostics in maintained stress tooling; no committed standalone harness or evidence.
- Validate the resulting outbox experiment with one lane, no more than 15 minutes per sample, fresh-main baseline, and report its URL, inputs and verdict before any repeat. Follow the existing INCONCLUSIVE/REGRESSION policy.

This is a prerequisite for #3232's maintained loaded investigation. It does not resolve the 112 B per-save EF notification allocation difference or establish acceptance against the historical pre-notification baseline.

### PR DESCRIPTION
**Pull request by thomhurst**
**Loaded verdict: INCONCLUSIVE. Do not merge on current CI or worker-retention correctness alone.**

[Run 34642337592](https://github.com/thomhurst/Dekaf/actions/runs/34642337592) completed the declared `outbox-1b` experiment: cheap exact-SHA A1/B/A2, five minutes per measured sample (75 seconds idle plus 225 seconds active), 180 seconds warmup per phase, 32 retained workers, 1,000-byte messages, profiling/fetch diagnostics off, and no second candidate or full run. Baseline is `1f28994b63b78b313b6dc8fe38cec52a17523734`; candidate/harness is `e1117501dc0d9db584c1951a3b0979d1bf381129`.

The worker-lifetime correction works in the loaded experiment: all five processes report requested/minimum/maximum/live workers as 32 with idle timeout -1, and every retained runtime sample from both preflights and A1/B/A2 remains at 32. All five workloads complete with committed messages equal to unique consumption and zero duplicate publications. A1/B/A2 deliver 1,051,025 / 955,013 / 1,016,470 unique messages.

Steady-state coverage is validated, but trend acceptance is INCONCLUSIVE. All three measured idle phases fail the 10% CPU trend check (first/last-third changes -12.87%, -16.49%, -18.16%). Candidate B and control A2 also breach the harness throughput slope/drift checks. The comparison retains these adverse aggregate rows:

- Delivered throughput: A1/B/A2 = 4643.989/4201.796/4487.867 msg/s; candidate differences -9.52%/-6.37%. The row is REGRESSION.
- Latency p50: A1/B/A2 = 375.085/417.245/380.115 ms; candidate differences +11.24%/+9.77%. The row is REGRESSION.

The overall comparison is INCONCLUSIVE because startup/trend assessment did not establish steady state; these regression rows are not waived or reported as a pass. Product source and root build inputs are byte-identical between the measured revisions, while the declared workflow overlays candidate stress sources onto both controls. This confirms the comparison setup, not harmlessness of the observed losses. No tolerances or samples were changed, and no repeat or additional paid experiment has been dispatched.

Original controls/candidate results, job log, comparison and warmup reports, all five process snapshots, and the 306 managed output files verified against the runner's SHA-256 manifest are retained at `D:/Dekaf-evidence/pr-3277-loaded-34642337592`. Pinned Git source archives and a verified reconstruction of the declared baseline harness overlay are included; that reconstruction is explicitly distinguished from a source copy retrieved from the runner. [Job comparison and warmup table](https://github.com/thomhurst/Dekaf/actions/runs/34642337592/job/103405218003).

Current remote head `5e57999fa37304a78b84dd4236b0034a50974bb1` has 29 terminal passing/neutral/skipped checks and clear subsequent review. The required fresh-main rebase is prepared locally as `18611ffaa56b90127e191033ce105723ce98edfe` over `da840f694e75beb14f962aebdfdeeab67494f7d0`, but is not pushed while this loaded result is unresolved. It changes only the three review-action digest pins; 342 Python tests pass with 12 platform skips, artifact/diff checks pass, and exact source/runtime evidence is verified at `D:/Dekaf-evidence/pr-3277-fresh-main`. It is not a fix for the remaining workload trends. Historical allocation/CPU requirements recorded in #3232 remain outside this worker-lifetime correction.

Earlier implementation and validation details:

Current head: `5e57999fa37304a78b84dd4236b0034a50974bb1`. Review follow-up fixes early leading-zero rejection and exercises the complete dispatch selection path. Python validation: 342 passed, 12 platform skips; artifact policy and diff checks pass. Evidence: `C:/git/Dekaf-evidence/pr-3277/worker-input`, including verified source/runtime inventories. The original C# validation below applies to unchanged runner source. Loaded run 34642337592 remains on its original `e1117501dc0d9db584c1951a3b0979d1bf381129` candidate; verdict pending. New-head CI/review also remains pending.

`fixed_worker_threads=32` previously set only the .NET minimum and maximum limits. Workers still retired while idle, so the outbox and consumer experiments failed steady-state validation despite completing their workloads. On Linux with SDK 10.0.401/runtime 10.0.12, a controlled check started 32 workers and observed zero after 25 seconds idle while the limits remained 32/32. Disabling idle retirement retained all 32.

Explicit fixed-worker runs now disable idle retirement and prestart the requested workers before warmup. Startup verifies actual limits and live count, rejects incomplete configuration, and writes per-process provenance beside the results. Both pinned controls and candidate receive identical setup. Input `0` preserves runtime defaults; decimal inputs with leading zeros are normalized before Bash hexadecimal conversion. The maintained `check-worker-pool` command verifies the pool across a 25-second idle interval without Kafka.

All synchronization, queued callbacks and provenance allocation occur once per process before warmup. Message loops gain no work. Retained worker stacks are part of the explicitly requested fixed-pool experiment. Existing duration, warmup, correctness and steady-state checks remain intact. No product code, gate tolerances or workload defaults change.

Validation on `e1117501dc0d9db584c1951a3b0979d1bf381129`, one commit over `1f28994b63b78b313b6dc8fe38cec52a17523734`:

- Full Release solution build: zero warnings/errors.
- All 69 stress-tool TUnit tests pass, including a process-isolated idle-retention check and invalid-configuration cases.
- The actual workflow Bash block reproduces three failures before the fix and passes normal, boundary, leading-zero and disabled inputs afterward.
- Full Python suite: 339 passed, 13 skipped; the skipped jq workflow case passes separately with jq available, leaving only 12 platform skips unexecuted.
- The final maintained runner on Linux prestarts and retains all 32 workers through 25 seconds idle. Artifact policy and diff checks pass. The final diff was reviewed for reuse, lifecycle safety and simplicity.

Exact final source, 475 loaded binary/report entries, raw logs, Linux output and image identity are SHA-256 verified outside removable worktrees at `D:/Dekaf-evidence/issue-3276`. Causal runtime/probe evidence remains at `D:/Dekaf-evidence/issue-3232/threadpool-probe`, including the complete Linux runtime and the explicit retention limitations of the initial configuration-only probe. The final validation archive is self-contained for its tested source and managed binaries; it references that retained Linux runtime/image.

Loaded acceptance is pending. The prior [outbox run](https://github.com/thomhurst/Dekaf/pull/3265#issuecomment-5638860580) and [consumer run](https://github.com/thomhurst/Dekaf/actions/runs/34637728473) remain INCONCLUSIVE. One corrected `outbox-1b` experiment will use cheap shape, five minutes per A1/B/A2 sample, 180 seconds warmup per phase, 32 workers, 1,000-byte messages and profiling off, with the fresh-main baseline above. This validates the corrected experiment setup; #3232 records the unresolved 112 B per-save notification difference and historical loaded acceptance. Do not merge before the loaded verdict and new CI/review cycle are assessed.

Closes #3276

## Summary by CodeRabbit

* **New Features**
  * Added an optional fixed worker-thread configuration for stress-test workloads.
  * Added validation that configured workers remain available after an idle period.
  * Added worker-pool status snapshots and clearer command-line help.
  * Added support for validating worker-pool settings before running experiments.

* **Documentation**
  * Documented fixed worker-thread configuration, warmup behavior, idle-retirement settings, and recorded worker metrics.

* **Tests**
  * Added coverage for valid, invalid, and boundary worker-count configurations.

### CODE REVIEW

**Comment by thomhurst** [comment]
Dispatched [run 34642337592](https://github.com/thomhurst/Dekaf/actions/runs/34642337592): `outbox-1b`, cheap exact-SHA A1/B/A2, five minutes per measured sample, 180-second warmup per idle/active phase, 32 prestarted retained workers, 1,000-byte messages, profiling/fetch diagnostics off and Dekaf only. Baseline `1f28994b63b78b313b6dc8fe38cec52a17523734`; candidate `e1117501dc0d9db584c1951a3b0979d1bf381129`. **Verdict pending.** No second candidate or full run is requested.

This experiment applies the diagnosed worker-lifetime correction equally to both controls and candidate. It does not resolve #3232's historical allocation/CPU requirements. The prior outbox and consumer runs remain INCONCLUSIVE with their original results preserved. No repeat will be dispatched before this run's terminal verdict is recorded and assessed.

File: .github/workflows/stress-tests.yml [hunk 1]
```diff
@@ -479,16 +479,21 @@ jobs:
         env:
           FIXED_WORKER_THREADS: ${{ inputs.fixed_worker_threads }}
         run: |
+          # Normalize decimal input before printf, which otherwise treats leading zeros as octal.
+          worker_count=$((10#$FIXED_WORKER_THREADS))
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply .github/workflows/stress-tests.yml:489]: Fixed in `5e57999fa37304a78b84dd4236b0034a50974bb1`. Initial selection now validates decimal digits, normalizes leading zeros, and then checks the 0–1024 range and non-publishing exact-SHA restriction. Invalid expressions and oversized integers are rejected before Bash arithmetic.

The regression test now executes the complete selection script, including real jq matrix generation and the timeout calculator, before the worker configuration script. It verifies nine valid counts (including `032`, `0000`, `08`, and padded boundary values), twelve invalid inputs, and ten disabled/publishing combinations. The original validation fails six valid-count cases; the fix passes. The complete Python suite passes 342 tests, with 12 existing platform skips.

Exact scripts, original/final source archives, raw red/green reports, Python/Bash/jq binaries and a 162-file source/runtime inventory verified against originals are retained at `C:/git/Dekaf-evidence/pr-3277/worker-input`. The final source archive was also verified against fresh Git output. Artifact policy and diff checks pass; the follow-up was reviewed for simplicity and scope.

This follow-up changes only dispatch validation and its tests. The C# runner, worker setup script and the existing `32` experiment configuration are unchanged. Run [34642337592](https://github.com/thomhurst/Dekaf/actions/runs/34642337592) remains pinned to its original `e1117501dc0d9db584c1951a3b0979d1bf381129` candidate and `1f28994b63b78b313b6dc8fe38cec52a17523734` baseline; its verdict remains pending. No additional paid run was dispatched. Review threads remain open for the subsequent bot/CI cycle.

[the remark this answers is not in the document]
Comment (thomhurst) [reply .github/workflows/stress-tests.yml:489]: Addressed by 5e57999fa37304a78b84dd4236b0034a50974bb1. The test now executes complete lane selection with real jq and timeout calculation before configuring the worker pool. Decimal normalization precedes bounds and publishing checks. The detailed validation and evidence disposition is in the other review reply.

**Comment by thomhurst** [comment]
**Loaded verdict: INCONCLUSIVE. Do not merge on current CI or worker-retention correctness alone.**

[Run 34642337592](https://github.com/thomhurst/Dekaf/actions/runs/34642337592) completed the declared `outbox-1b` experiment: cheap exact-SHA A1/B/A2, five minutes per measured sample (75 seconds idle plus 225 seconds active), 180 seconds warmup per phase, 32 retained workers, 1,000-byte messages, profiling/fetch diagnostics off, and no second candidate or full run. Baseline is `1f28994b63b78b313b6dc8fe38cec52a17523734`; candidate/harness is `e1117501dc0d9db584c1951a3b0979d1bf381129`.

The worker-lifetime correction works in the loaded experiment: all five processes report requested/minimum/maximum/live workers as 32 with idle timeout -1, and every retained runtime sample from both preflights and A1/B/A2 remains at 32. All five workloads complete with committed messages equal to unique consumption and zero duplicate publications. A1/B/A2 deliver 1,051,025 / 955,013 / 1,016,470 unique messages.

Steady-state coverage is validated, but trend acceptance is INCONCLUSIVE. All three measured idle phases fail the 10% CPU trend check (first/last-third changes -12.87%, -16.49%, -18.16%). Candidate B and control A2 also breach the harness throughput slope/drift checks. The comparison retains these adverse aggregate rows:

- Delivered throughput: A1/B/A2 = 4643.989/4201.796/4487.867 msg/s; candidate differences -9.52%/-6.37%. The row is REGRESSION.
- Latency p50: A1/B/A2 = 375.085/417.245/380.115 ms; candidate differences +11.24%/+9.77%. The row is REGRESSION.

The overall comparison is INCONCLUSIVE because startup/trend assessment did not establish steady state; these regression rows are not waived or reported as a pass. Product source and root build inputs are byte-identical between the measured revisions, while the declared workflow overlays candidate stress sources onto both controls. This confirms the comparison setup, not harmlessness of the observed losses. No tolerances or samples were changed, and no repeat or additional paid experiment has been dispatched.

Original controls/candidate results, job log, comparison and warmup reports, all five process snapshots, and the 306 managed output files verified against the runner's SHA-256 manifest are retained at `D:/Dekaf-evidence/pr-3277-loaded-34642337592`. Pinned Git source archives and a verified reconstruction of the declared baseline harness overlay are included; that reconstruction is explicitly distinguished from a source copy retrieved from the runner. [Job comparison and warmup table](https://github.com/thomhurst/Dekaf/actions/runs/34642337592/job/103405218003).

Current remote head `5e57999fa37304a78b84dd4236b0034a50974bb1` has 29 terminal passing/neutral/skipped checks and clear subsequent review. The required fresh-main rebase is prepared locally as `18611ffaa56b90127e191033ce105723ce98edfe` over `da840f694e75beb14f962aebdfdeeab67494f7d0`, but is not pushed while this loaded result is unresolved. It changes only the three review-action digest pins; 342 Python tests pass with 12 platform skips, artifact/diff checks pass, and exact source/runtime evidence is verified at `D:/Dekaf-evidence/pr-3277-fresh-main`. It is not a fix for the remaining workload trends. Historical allocation/CPU requirements recorded in #3232 remain outside this worker-lifetime correction.

**Comment by thomhurst** [comment]
PR #3277 was merged as `689e1bb499963c6445893928231c0c21fd396080` after the INCONCLUSIVE result was recorded above. The merged tree is byte-identical to the locally prepared fresh-main commit `18611ffaa56b90127e191033ce105723ce98edfe`; that local commit was not pushed. The merge is not treated as a PASS for the loaded run or as approval of the recorded aggregate losses.

Additional read-only analysis uses exactly the idle windows selected by `stress_warmup.py`, without removing any observations. Each A1/B/A2 first window contains three lease acquisitions across approximately 24 seconds; each last window contains two across approximately 23 seconds. Probe counts are 27/28/29 early versus 24 late. Corresponding CPU rates are 8.928/9.039/8.888 ms/s early and 7.779/7.549/7.274 ms/s late. This confirms unequal periodic work in the short trend windows; it does not attribute every CPU difference or resolve the separate active-throughput findings. The original verdict and every regression row remain unchanged.

The exact window inputs and analysis are retained beside the original artifacts in `D:/Dekaf-evidence/pr-3277-loaded-34642337592`. No additional outbox run, tolerance change, sample exclusion or product change follows this observation. The fixed-worker implementation is on main; the historical outbox CPU/allocation and steady-state acceptance questions remain unresolved in the recorded evidence.
